### DISCUSSION TITLE
What is the canonical way to modify a pipeline's language data?

### DISCUSSION THREAD

**Discussion by afparsons** [explosion/spaCy#7477] [Help: Coding & Implementations] [answered]
**Goal:** I wish to modify how tokens are prescribed the `LIKE_NUM` attribute.

**Context**
---

I've defined a custom `lex_attrs.py` file (akin to [this](https://github.com/explosion/spaCy/blob/master/spacy/lang/en/lex_attrs.py)), namely to:
- use my own trigger lists (I have replaced/supplemented `_num_words` and `_ordinal_words`; for example, adding _half_, _dozen_, and _score_)
- define my own `like_num()` function

**Question**
---

Is this acceptable, or a hack considered harmful?

```python
# abbreviated imports (eliding obvious ones)
from haleynlp.languages.en.language_data.lex_attrs import LEX_ATTRS

# abbreviated support function
def analyze_tokens(doclike, *attributes):
    """
    Iterates through tokens, calls attributes, and returns a transformed Pandas DataFrame
    """
    ...

English.Defaults.lex_attr_getters = LEX_ATTRS
nlp = spacy.load('en_core_web_md')
doc = nlp('One half dozen and four score.')
analyze_tokens(doc, 'like_num')


# output from analyze_tokens(); note that thanks to my custom lex_attr, half, dozen, and score are considered LIKE_NUM
             0     1      2      3     4      5      6
text       One  half  dozen    and  four  score      .
like_num  True  True   True  False  True   True  False
```

As an aside, I do not understand why the following does not work (well, it only works after the model is reloaded). Is it because attaching LEX_ATTRS goes to the _Language_ and not the _pipeline_?

```python
>>> nlp = spacy.load('en_core_web_md')
>>> nlp.Defaults.lex_attr_getters = LEX_ATTRS

>>> doc = nlp('One half dozen and four score.')
>>> analyze_tokens(doc, 'like_num')

# half, dozen, and score return False for token.like_num

>>> nlp = spacy.load('en_core_web_md')
>>> doc = nlp('One half dozen and four score.')
>>> analyze_tokens(doc, 'like_num')

# suddenly it works
```

If `English.Defaults.lex_attr_getters = LEX_ATTRS` is **_not_** considered a good practice, what is the canonical way to achieve what I am trying to do? Bear in mind that I also will be adding pipeline components (re [this](https://github.com/explosion/spaCy/discussions/7430)) and eventually saving and distributing this pipeline.

**Addenda**
---

Apologies in advance for my naivety; I originally thought I could accomplish everything I wanted with the `Matcher()` and some  custom components, but I'm now finding I might have to embark on more advanced usage.

**Feedback:** spaCy has great documentation, but the [Language Data](https://spacy.io/usage/linguistic-features#language-data) section feels a little sparse.

**Errata**
---

In the course of writing this post, I drafted some statements which were either incorrect or irrelevant. I've included them below to demonstrate my state of understanding.

~~As I understand, there is no way for me to "insert" this new ruleset into a pre-trained pipeline. That is, I cannot take `nlp = spacy.load(en_core_web_md)` and somehow force my new `lex_attrs.py` file into it~~

_This isn't true. I've figured out how to modify `.Defaults.lex_attrs_getter`, even if it is not considered a good practice._

_Irrelevant:_

Likewise, I cannot copy the components from one pipeline into another. I know this because:
1. my naive attempt at doing so throws errors (where `CustomEnglish` is a subclass per the [documentation](https://spacy.io/usage/linguistic-features#language-subclass)):
    ```
    nlp = CustomEnglish()
    default = spacy.load('en_core_web_md')
    for component in ('tok2vec', 'tagger', 'parser', 'attribute_ruler', 'lemmatizer', 'ner', ):
        nlp.add_pipe(component, source=default)

    [E945] Can't copy pipeline component 'tok2vec' from source. Expected loaded nlp object, but got: <class 
    'spacy.lang.en.English'>
    ```

2. the [documentation says so](https://spacy.io/api/language#add_pipe) (I think?)
    > Make sure that the vocab, vectors and settings of the source pipeline match the target pipeline.

**Comment by adrianeboyd** [comment, marked as answer]
On modifying the language defaults:

The canonical way to modify these settings in a config is to use the `nlp.before_creation` callback, which lets you modify the `Defaults` **before** the pipeline is initialized with `English()`. It's the config version of what you're doing when modifying `English.Defaults` before loading. It looks like the examples in this section: https://spacy.io/usage/training#custom-code-nlp-callbacks

You can add a `before_creation` callback to an existing pipeline like `en_core_web_md` by creating a custom model package with the callback inserted into the config and the callback code packaged as part of the model. Then people using the model just need to run `spacy.load("en_my_custom_model")` and they get all the settings loaded in without any additional steps. There are a lot of small steps to create the new model package, but it's not hard to do. You would:

1. Save a local copy of `en_core_web_md` with `nlp.to_disk()`
2. Update the metadata in `meta.json` to change the model name
3. Write a callback modifying the language defaults in a new file called `functions.py` (or whatever filename you want)
4. Add the callback to the `config.cfg` as `nlp.before_creation`
5. Use `spacy package model_dir --code functions.py` to package it all as a pip-installable package that works just like `en_core_web_md`

Note that this method modifies the `English.Defaults` for all other English pipelines loaded in the same script, too. Depending on the order in which pipelines are loaded and used, you might also get weird side effects due to the lexeme cache, which could lead to fairly confusing behavior. For example, if you load and run `nlp1` on some texts but haven't seen all the "like num" words before you load `nlp2` with a custom `like_num` definition, then you get different behavior for lexemes added to `nlp1` afterwards. This isn't a problem that comes up much in typical usage, though.

If you want to avoid these side effects, then you'd need to use the `CustomEnglish` class option instead. With this method, if you're modifying `en_core_web_md`, you'd edit `nlp.lang` in `config.cfg` rather than adding a callback. You'd still need to package your custom language code with `--code`.

Actually, thinking about the side effects, the custom language class is a cleaner solution overall. In general, it's way more of a practical concern if you're overriding attributes that are used by the statistical components (`PREFIX`, `SUFFIX`, `SHAPE`) and it's much less likely to break someone else's code using `en_core_web_sm` in parallel if `LIKE_NUM` is slightly different than expected. Most people wouldn't notice at all because they never use `LIKE_NUM` and they only load one model in a given script.

I think we recommended modifying the `English.Defaults` because it was much easier than creating a custom language in v2, but now that it's about the same amount of effort, I think we could reconsider our recommendations here.

More detail about the language defaults is here: https://spacy.io/api/language#defaults

----------

On sourcing components with custom language classes:

The checks for compatibility for `source` are still a bit under development. There aren't any checks for vectors, etc. yet so you can potentially end up with a broken combination of components. The only check that's implemented restricts you to sourcing from the exact same language class (`English` source for `English` pipeline) and it shows this confusing error message otherwise, when it should give a different error message if only mismatched languages are the problem. It should probably allow subclasses like `CustomEnglish` when sourcing `English` components. This kind of behavior with `CustomEnglish` is a bug, we'll look into it.

**Reply by afparsons** [reply]
I had to put this project down for a few days and work on more pressing matters. My apologies for not responding and/or marking this as answered. I'm glad @svlandeg went ahead and did so already.

Thank you @adrianeboyd for the response. As always, spaCy's/Explosion's support is fantastic :+1: 

---

> weird side effects due to the lexeme cache, which could lead to fairly confusing behavior

Ah-ha, I think that might explain what I described in the original post. 

---

> There are a lot of small steps to create the new model package . . .  then you'd need to use the CustomEnglish class option instead.

For future passers-by, here's what to do (as of `spaCy 3.0.3`):

1. create and register a custom language per "[Creating a custom language subclass](https://spacy.io/usage/linguistic-features#language-subclass)"
    - we'll give this language a very original name: `custom__en` 
2. save `en_core_web_md` to disk with `nlp.to_disk()`
3. in `meta.json`:
    - change: `"name":"core_web_md"` to `"name":"custom__core_web_md"`
    - do not change: `"lang":"en"` ([because this is an ISO 639-1 code](https://spacy.io/api/data-formats#meta))
4. in `config.cfg`:
    - change: `[nlp]lang = "en"` to `[nlp]lang = "custom__en"`
5. Package everything with `spacy package . . . --code /path/to/my/custom/language.py`. I suggest writing a shell script for future fast iteration, but this is, of course, workflow-dependent.
6. (shell) `(venv) $ cd </path/to/output> && python -m pip install .`
7. (Python) `nlp = spacy.load('en_custom__core_web_md')` 

Brilliant. Now my `nlp` object is using the custom language data (in my case, just a custom `lex_attrs` at the moment).
